lxd/images.go

```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
 }
 
 func autoUpdateImages(ctx context.Context, d *Daemon) error {
-	projectNames := []string{}
+	imageMap := make(map[string][]db.Image)
+
 	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
-		projects, err := tx.GetProjects(db.ProjectFilter{})
+		var err error
+
+		images, err := tx.GetImages(db.ImageFilter{AutoUpdate: true})
 		if err != nil {
 			return err
 		}
 
-		for _, project := range projects {
-			projectNames = append(projectNames, project.Name)
+		for _, image := range images {
+			imageMap[image.Fingerprint] = append(imageMap[image.Fingerprint], image)
 		}
 
 		return nil
 	})
 	if err != nil {
-		return errors.Wrap(err, "Unable to retrieve project names")
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Debug("Error getting nodes", log.Ctx{"fingerprint": fingerprint})
+			continue
+		}
+
+		var nodeIDs []int64
+
+		for _, node := range nodes {
+			err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+				var err error
+
+				nodeInfo, err := tx.GetNodeByAddress(node)
+				if err != nil {
+					return err
+				}
+
+				nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+				return nil
+			})
+			if err != nil {
+				return errors.Wrap(err, "Unable to retrieve cluster node information")
+			}
+		}
+
+		// If multiple nodes have the image, select one to deal with it.
+		if len(nodeIDs) > 1 {
+			selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
+			if err != nil {
+				logger.Debug("Failed to get stable random value", log.Ctx{"err": err})
+				continue
+			}
+
+			if d.cluster.GetNodeID() != selectedNode {
```

Comment:
@monstermunchkin maybe a little comment here explaining that the matching node will run the task and explain that this is to stop multiple nodes trying to update and distribute the image concurrently.